src/main/java/com/google/collegeplanner/servlets/SavePlanServlet.java

```diff
@@ -0,0 +1,183 @@
+import com.google.api.client.googleapis.auth.oauth2.GoogleIdTokenVerifier;
+import com.google.api.client.json.gson.GsonFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.collegeplanner.data.Plan;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+
+/** Servlet that saves and returns a users plans.*/
+@WebServlet("/api/planner/save")
+public class SavePlanServlet extends BaseServlet {
+  /**
+   * Verifier for Google OAuth that validates tokens and ensures the source of the token is
+   * acceptable according to the client ID.
+   */
+  GoogleIdTokenVerifier verifier;
+  DatastoreService datastore;
+  /**
+   * Public identifier for this app to use Google OAuth 2.0. It is found in cloud console
+   * credentials.
+   */
+  final static String clientId =
+      "267429534228-vvsi2uldmpji3rgs1qd3a41rceciaaaq.apps.googleusercontent.com";
+
+  public SavePlanServlet() {
+    this(DatastoreServiceFactory.getDatastoreService(),
+        new GoogleIdTokenVerifier.Builder(UrlFetchTransport.getDefaultInstance(), new GsonFactory())
+            .setAudience(Collections.singletonList(clientId))
+            .build());
+  }
+
+  public SavePlanServlet(DatastoreService datastore, GoogleIdTokenVerifier verifier) {
+    this.verifier = verifier;
+    this.datastore = datastore;
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    GoogleIdToken idToken;
+    try {
+      idToken = verifier.verify(request.getParameter("idToken"));
+    } catch (GeneralSecurityException e) {
+      respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+      return;
+    }
+
+    if (idToken == null) {
+      respondWithError("Invalid user.", HttpServletResponse.SC_UNAUTHORIZED, response);
+      return;
+    }
+    ArrayList<Plan> plans = getPlans(idToken, response);
+    if (plans == null) {
+      return; // Error has already been written.
```

Comment:
Instead of hoping that the error has already been written, I would recommend writing the error here, or throwing an error in getPlans instead. It makes it easier to enforce that the proper responses are being written to the client if it generally all happens in one place.

also, need two spaces before comment